Dark theme fixes

cryham opened the pull request:
Related issues:
#5489 Dark Theme, #6275 High Contrast, #1947

## Fixes

Here are some (most) fixes I did to make GitExt usable in my black-violet theme.
GitExt is not black / white theme aware, as it uses some fixed colors set in code. Even worse, some are just for background without matching text set too.
Either way I think I fixed most of the stuff I saw (in Browse and Commit dialogs).

## Proposed changes

There is still a remaining issue with text / code editor line numbers background.
I just hid it in my [fork here](https://github.com/cryham/gitextensions/commit/f29130419f11fcb025e5cc8d2c4667def6205c0f), for now (in internal viewer only).

Second thing left, is that changing themes black / white doesn't change GitExt colors in setting (Appearance). I think it'd  be great to have 2 columns with colors (twice as many): one for white, other for dark themes.

## Screenshots

### Before

![Before](https://i.imgur.com/O1NgKe6.png)

### After

![After](https://i.imgur.com/0eJ0jWP.png)

## Test environment(s) <!-- Remove any that don't apply -->

- GIT version 2.21
- Windows 7 SP1

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented:
Can you please open a separate PR to sign contributors.txt?

RussKie commented:
Sexy!
We'll need to test in the light mode, but it is an awesome start!

gerhardol commented:
The dark mode definitely look better now...

Stupid question: How do I activate the dark mode?
Win10 1809 with Dark theme by default.

The light theme is changed, see the screenshot. (Not considered what I prefer.)

![image](https://user-images.githubusercontent.com/6248932/58290994-02742300-7dbc-11e9-8254-235e485531c6.png)

cryham commented:
Right.
So the dark theme is detected and set automatically in code when system theme color luminosity is 0.5 or below.
Not sure about Win 10, I don't have access to it yet.
In Win 7 I test by switching my theme 
I signed contributors.txt in #6659

gerhardol commented:
Win10 dark mode is not activating, will test with other themes

It would be good if the mode could be overridden too, i.e. default/light/dark.
(if one column is kept, this could control what is updated too).

I would be surprised if this PR is the complete solution for a dark theme, but it for sure is a great improvement.

gerhardol commented:
First, I am not a 'themer' myself. My wife does not like bright screens though...
This is something others have asked for, so it is a nice addition to GE.

Some background regarding Win10 dark theme:
Dark mode in Win10 only affect UWP apps, no simple way to find if this setting in other apps.
You can use HighContrast-Dark, that is very ugly...
In Win7 it was easier to apply custom themes. In Win10, this is harder, something like WindowsBlinds can be used:
https://www.howtogeek.com/222614/how-to-enable-windows-10s-hidden-dark-theme/

Using system colors consistyenly changes the normal theme too. Thisreverts some of the careful changes that @drewnoakes applied but I am mostly fine with that. Tweaks will have to be done.

This is what GE looks with the DarkMode skin.
Some items are not applied for me.

![image](https://user-images.githubusercontent.com/6248932/58357323-a15f5480-7e7a-11e9-9e50-480937c34e64.png)

RussKie commented:
Thank you Gerhard, that doesn't look too bad - looks like few more tweaks are required

cryham commented:
Fixed line numbers on Diff tab. They go dark with theme.
I also changed graph lane colors somewhat. There was one too dark color. I generally like them more now.

File Tree tab is a different story. For dark theme only some file types are dark (e.g. txt, md, gitignore). All language files have files here \Externals\ICSharpCode.TextEditor\Project\Resources\C#.xshd. Lots of them but the top section <Environment> seems to look almost identical across them. And yeah <Default color="Black" bgcolor="#FFFFFF"/> is white background (bad for eyes in dark theme.
Even worse, all those files go into ICSharpCode.TextEditor.dll as resources.

How would you approach it? There are themes for Console, I guess it'd be great to have them for File Tree viewer too. But this seems like a big task to me.

gerhardol commented:
> Should there be 2 sets, i.e. a 2nd column for all colors for dark theme?

With the approach in this PR, with SystemColors and special handling depending on light/dark only, I would first like to have "hardcoded" translation. If this is futile, I prefer a second set of colors to avoid a very manual setup.

The discussions in #5489 was about creating theme_s_ by a proxy. 
This PR uses the light variant by choosing to use SystemColors where possible and choosing a light/dark variant where needed. There are quite many other forms that will need adjustments still with hardcoded colors. With certain dark themes, GE may even look nice. Use of SystemColors only also avoids the issue where some controls uses SystemColors internally.

What do the prominent WinForm developers think?
@NikolayXHD @persn @digitalcoyote 

> Even worse, all those files go into ICSharpCode.TextEditor.dll as resources.

GE uses the text editor in the submodule, it can be changed.

Some issues I noted when using Win10 dark theme. All issues in all windows themes will not be solved now....

Browse
Icon not seen
![image](https://user-images.githubusercontent.com/6248932/58371882-087e1700-7f16-11e9-9f3c-6f961b4bb4b1.png)

![image](https://user-images.githubusercontent.com/6248932/58371970-05375b00-7f17-11e9-8508-e30ff4dc30e2.png)

Commit
![image](https://user-images.githubusercontent.com/6248932/58371866-cead1080-7f15-11e9-8626-1a274f0e8c29.png)

![image](https://user-images.githubusercontent.com/6248932/58371873-e5536780-7f15-11e9-90b5-fc4708b26d86.png)

About
Maybe link color in theme?
![image](https://user-images.githubusercontent.com/6248932/58371903-44b17780-7f16-11e9-9d3b-3d918ef013c8.png)

GitHub
![image](https://user-images.githubusercontent.com/6248932/58371913-6f033500-7f16-11e9-9472-12578667a2c0.png)
![image](https://user-images.githubusercontent.com/6248932/58371925-8cd09a00-7f16-11e9-8d3b-2a89d5e74a0e.png)

Other
![image](https://user-images.githubusercontent.com/6248932/58371983-2730dd80-7f17-11e9-9eee-5378520e139c.png)

![image](https://user-images.githubusercontent.com/6248932/58371992-40398e80-7f17-11e9-901a-b0d1fd6bbebb.png)

RussKie commented:
we will need to have a second set of icons for the dark theme, let's leave
it out for now

On Sun, May 26, 2019, 6:52 AM Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
>
> For light theme, this affects two backdrops in FormBrowse.Designer.cs and
> RevisionGraphLaneColor.cs
> Both changes are fine for me
> The rest is just using SystemColors and some light/dark awareness required
> for dark mode.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXXTPLWPLKEHW3XZ7FTPXGRKFA5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZWNZII#pullrequestreview-242015393>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXV7QCCTQO2A2BW6E5LPXGRKFANCNFSM4HPKTI3A>
> .
>

cryham commented:
Icons that are black (or almost) and aren't visible are (inside GitUI\Resources\Icons):

Delete.png
Develop.png
FileStatusRenamed.png
github.png
GitHub32.png
Link.png
PullRequest.png
RecoverLostObjects.png
ShowWhitespace.png
Translate.png
WhitespaceIgnore.png
WhitespaceIgnoreAll.png
WhitespaceIgnoreEol.png

I could probably make a light gray version of them and if we can switch in code at init between them that would work right?

cryham commented:
OK I think the icons switch are done now.
I didn't know how to check few (e.g. pull request, delete is red?). I left github as it is.

gerhardol commented:
GE looks mostly OK in Win7 with your theme.
Not much worse than the diff margin (as described above), the default diff, hlight colors and the following from Apply Patch (random check though):
![image](https://user-images.githubusercontent.com/6248932/58498456-60b75200-817e-11e9-966f-a55a2667e2c9.png)

However in Win10 with the dark mode I test with, there are some artifacts (mostly black text or links on dark/black background). Most of that are due to hardcoded colors, like Color.Black.
Do you want a PR for that (to your branch), me to push to your branch directly or fix it yourself?

When it is OK for me I will try to make a better review (I am away this weekend) (do not expect many comments blocking merge for me, but more about improvements). 
More users must test/review before merge though.

cryham commented:
So I think a PR to my branch then (if not too much trouble).
I think I searched for Black colors set in code but likely still didn't catch them all.
I don't have Win 10 yet and don't know GUI that well to find / show them.

Apropos this previous screen with disabled buttons on Apply Patch etc. Is there another way? I'm used to disabled things looking like that. I.e. barely visible or invisible in menus of other programs.

Just for info here, there are like 2 more details I see (that I don't know how or if possible to fix):
Here's a screen with them:
![screen](https://i.imgur.com/70OrD17.png)
1. Checkboxes (in menus) have black tick, which makes them invisible for me, only the rectangle around is slightly visible but not much (I think it's system selected color).
2. Some icons have still few pixels almost white on edges (around icon). E.g. Commit, GitExt in settings tree, Fonts etc. But I think this would be a lot of work to fix those pixels in icons, maybe easier to make new. Probably not worth it.

gerhardol commented:
For most of the black text on black I see: That seem to be the theme, for inactive buttons.
![image](https://user-images.githubusercontent.com/6248932/58515452-27460d00-81a5-11e9-8486-b4ceb443adf5.png)

There are some hardcoded use of Color.Black, but many of those are in ICSharpCode.TextEditor, that is a separate repo.

The links are the most difficult to read now. (other than disabled)

For the background for line number in diff: Should that really be black and not the same as the file viewer?

![image](https://user-images.githubusercontent.com/6248932/58516091-22825880-81a7-11e9-9185-fc05ad4d9a80.png)

Otherwise most artifacts has been fixed. No PR from me today at least, probably few changes if any at all tomorrow.

todo?
AsIs, Special handling (hardcoded) or double configuration for special colors like diff?
Fix TextEditor now too?
Replace SystemColors with "proxy" to prepare for theming within GE?

RussKie commented:
Is it possible to make text lighter? Or is it a standard font color in the theme?

> AsIs, Special handling (hardcoded) or double configuration for special colors like diff?
> Fix TextEditor now too?
> Replace SystemColors with "proxy" to prepare for theming within GE?

We can open separate issue for these.
I don't expect _all_ things to be addressed in one go, I fully appreciate amount of effort required.

gerhardol commented:
> Is it possible to make text lighter? Or is it a standard font color in the theme?

Set in the Windows theme, so a bad theme will give a bad experience.
This a reason to have a "proxy"/theme engine to override certain colors, not just the colors that clash with the theme (like the diff colors in my example). That could be done separately). Reason to do it here is to reduce noise, to avoid 

This PR makes GE usable in dark theme and high contrast, more changes to make the app great indark themes.

gerhardol commented:
For links: The colors are fixed in WinForms (W3C standard, derived from InternetExplorer).
Most forms use the default which is hard to read with dark themes.
A few forms (like Settings/Appearance) use SystemColors.HotTrack instead.

This PR makes a special change in About Form. That improves, but hard to do for all links.
Should maybe be reverted.

A solution is to set colors for all LinkLabel, another to use a custom control.

gerhardol commented:
For the line numbers background, the following looks better for me.

GE3.1

![image](https://user-images.githubusercontent.com/6248932/58595674-bcaed300-8271-11e9-97f3-ffbb9a1ed2dc.png)


This PR, Win10 default and the theme I test with

![image](https://user-images.githubusercontent.com/6248932/58595976-abb29180-8272-11e9-9551-d11554b098f0.png)
![image](https://user-images.githubusercontent.com/6248932/58596108-195ebd80-8273-11e9-89f8-fe86ee006268.png)

Win10 patch
![image](https://user-images.githubusercontent.com/6248932/58596014-ce44aa80-8272-11e9-8f07-912404758b79.png)
![image](https://user-images.githubusercontent.com/6248932/58596076-f9c79500-8272-11e9-9c0e-7209c9169a93.png)

cryham commented:
OK. I think I'll do the 2nd column for Diff colors next. Will be convenient when switching system theme, if they change automatically.
Only thing with line numbers background I'm worried about is finding good colors, for text too. So they don't get too much or too little contrast on various black/gray/white themes.

RussKie commented:
May be look at how our competitors do it.
GitKraken and Fork both have dark colored themes.
I can post few screenshots from Fork, if you want.

On Thu, May 30, 2019, 5:41 PM Crystal Hammer <notifications@github.com>
wrote:

> OK. I think I'll do the 2nd column for Diff colors next. Will be
> convenient when switching system theme, if they change automatically.
> Only thing with line numbers background I'm worried about is finding good
> colors, for text too. So they don't get too much or too little contrast on
> various black/gray/white themes.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXSQNSQW4DOMNQNTNA3PX6AMNA5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWRT6BI#issuecomment-497237765>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXVKQNPMRHVTN6VYZDLPX6AMNANCNFSM4HPKTI3A>
> .
>

gerhardol commented:
> May be look at how our competitors do it.

This PR makes GE from being horrible to play nicely with the Windows theme, using SystemColors where appropriate and a adjusting a few situations to dark/light where GE mselects colors. It does not handle:
 * GE default colors (like diff colors that do not match the theme) 
  -> could have separate light/dark colors here, but await theming first
 * Theme colors clashing with the theme (like links dark blue)
  -> also a general question that will require override of the bad controls
 * Theme bad setup (like black text on dark background like disabled controls)
  -> Not much to do about
 * Some corner cases where bad hardcoded colors or SystemColors use give a bad layout. The line number background with the theme I test with is one example.
  ->Depends on the theme, will have to fixed where found

Most of the above should not be handled in this PR.

To have built-in themes will require a "theme-engine". That is a lot more work as it will require to proxy SystemColors and override controls that use SystemColors. This PR is a step to that as it identifies where colors are set.

gerhardol commented on GitUI/CommandsDialogs/FormAbout.cs:
```diff
@@ -21,6 +22,11 @@ public FormAbout()
 
             environmentInfo.SetCopyButtonTooltip(_copyTooltip.Text);
 
+            Color clrLink = ColorHelper.IsLightTheme() ? Color.Blue : Color.SkyBlue;
```
While the links are hard to read in dark themes, this is a general problem and this increases inconsistency.
Some links use SystemColors.HotTrack instead (see for instance CommitInfo. Looks a little worse.
Even if I commented on this earlier, I propose to revert this change, at least for now.

cryham replied:
Why not use HotTrack in About too then? It is skyblue on my theme and looks good.
I changed to blue it in next commits (they aren't that visible now).

gerhardol replied:
We could change all to HotTrack, but there are quite some LinkLabel uses.
It is maybe better to use a custom control to set also visited links.
The color should probably be unset, to use the defaults.

gerhardol commented on GitUI/Editor/FileViewerInternal.cs:
```diff
@@ -60,7 +60,10 @@ public FileViewerInternal()
                         TextEditor.ActiveTextAreaControl.TextArea.TextView.GetLogicalLine(e.Y)));
             };
 
-            HighlightingManager.Manager.DefaultHighlighting.SetColorFor("LineNumbers", new HighlightColor(Color.FromArgb(80, 0, 0, 0), Color.White, false, false));
```
The background looks like a visual artifact in the theme I test with and I suggest changing this (maybe as in https://github.com/gitextensions/gitextensions/pull/6651#issuecomment-497133999) also if GE normal theme changes.

cryham replied:
Right. I changed lines to just use System Control and Text colors. Looks good on my black/gray/white themes.

gerhardol commented on GitUI/UserControls/RevisionGrid/Graph/RevisionGraphLaneColor.cs:
```diff
@@ -8,15 +8,15 @@ namespace GitUI.UserControls.RevisionGrid.Graph
     public static class RevisionGraphLaneColor
     {
         internal static readonly IReadOnlyList<Color> PresetGraphColors = new[]
-{
```
Changed lane colors, OK for me

RussKie commented:
I'm too busy at the moment with the new job, leaving it with you.
I'm hoping Crystal isn't going to disappear immediately after the merge.

On Fri, May 31, 2019, 4:38 AM Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* approved this pull request.
>
> OK for me (maybe About Link).
>
> I would like to see more reviewers before merge, but if no one joins the
> next days I suggest this is merged and the details that are changed are
> adjusted after.
>
> Can you pleas rebase and squash?
> (I see no use of the specific commits here.)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXSHBFQJSUFKCEZYVTDPYANJBA5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2FIRJA#pullrequestreview-243959972>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRFHSYMEU5NEHVTW4LPYANJBANCNFSM4HPKTI3A>
> .
>

RussKie commented:
Totally agree. I don't expect this PR will solve all our theming problems.
It is a huge effort as such, and I massively appreciate it.

I was pointing that we can look up the colour selection our
more-theme-advanced competitors have made and take an inspiration from that.

On Thu, May 30, 2019, 8:57 PM Gerhard Olsson <notifications@github.com>
wrote:

> May be look at how our competitors do it.
>
> This PR makes GE from being horrible to play nicely with the Windows
> theme, using SystemColors where appropriate and a adjusting a few
> situations to dark/light where GE mselects colors. It does not handle:
>
>    - GE default colors (like diff colors that do not match the theme)
>    -> could have separate light/dark colors here, but await theming first
>    - Theme colors clashing with the theme (like links dark blue)
>    -> also a general question that will require override of the bad
>    controls
>    - Theme bad setup (like black text on dark background like disabled
>    controls)
>    -> Not much to do about
>    - Some corner cases where bad hardcoded colors or SystemColors use
>    give a bad layout. The line number background with the theme I test with is
>    one example.
>    ->Depends on the theme, will have to fixed where found
>
> Most of the above should not be handled in this PR.
>
> To have built-in themes will require a "theme-engine". That is a lot more
> work as it will require to proxy SystemColors and override controls that
> use SystemColors. This PR is a step to that as it identifies where colors
> are set.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXR7BEVYRZYTJRIQ6XLPX6XH3A5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWSA7TA#issuecomment-497291212>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXSZ47KUQD343Y4P3H3PX6XH3ANCNFSM4HPKTI3A>
> .
>

gerhardol commented:
I plan to merge this in the next days unless there are protests. If there are comments on the (minor) visual changes in normal themes, we handle them later.

@cryham 
Can you please rebase on gitextensions/master?

cryham commented:
I think I did just now, but that was my first time, so I'm not sure if this went bad or not.

pmiossec commented:
No, that's not good :disappointed:
I'm not in front of a computer so I can't really diagnose what you did but at least at the you did a merge.

What @gerhardol was expecting is that you rebase all your commits onto 'upstream/master'.

So, if you did it well, you should have seen **only** your commits in the 'commits' tabs of this PR in github.

cryham commented:
OK now it is good?

pmiossec commented:
I don't know if it contains all your changes in this commit, but in terms of git history, yes, it seems OK.

I think @gerhardol will do a last quick check but seems :+1:

gerhardol commented:
@cryham 
Thanks for the contribution

Your rebase had the same result as mine, so it should be correct...
We normally prefer rebase to avoid merge conflicts (better done by the PR submitter) and to get a cleaner history. I hope you learnt something!

If you are interested in improving the dark theme further, I suggest you open a new feature.

cryham commented:
Great. Yes, I did.

RussKie commented:
Thank you :+1:

On Wed, 5 Jun 2019 at 07:31, Crystal Hammer <notifications@github.com>
wrote:

> Great. Yes, I did.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXWHFNLE6ALIEQJ2IRDPY3NMTA5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW555WA#issuecomment-498851544>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRZRSFY6L4I7L7OUITPY3NMTANCNFSM4HPKTI3A>
> .
>

hinstw commented:
I saw this issue in change-log for version 3.2.0 but could not find a way to change UI theme. Can someone please tell me how to change theme, or is it not possible yet? (i assumed that it should be somehow possible because the issue title says "fixes" so it implies that the dark theme exists)

NikolayXHD commented:
The fixes here make GitExtension look better with  Windows dark themes.

gerhardol commented:
@hinstw 
Some hints to use a System UI dark theme in Win10
https://github.com/gitextensions/gitextensions/pull/6651#issuecomment-495798173

RussKie commented:
We will porting the app to .NET Core, which is also open sourced.
So theming related fixes can be sent that way 😉

On Mon, Sep 16, 2019, 7:50 PM Gerhard Olsson <notifications@github.com>
wrote:

> @hinstw <https://github.com/hinstw>
> Some hints to use a System UI dark theme in Win10
> #6651 (comment)
> <https://github.com/gitextensions/gitextensions/pull/6651#issuecomment-495798173>
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6651?email_source=notifications&email_token=ABBTEXQ4FZ3RPOMEGRBQ5VDQJ62MLA5CNFSM4HPKTI3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6ZZC6A#issuecomment-531861880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXRI5YMP6PHENNE7P43QJ62MLANCNFSM4HPKTI3A>
> .
>
